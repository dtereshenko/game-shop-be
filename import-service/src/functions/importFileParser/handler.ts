import { S3Handler } from "aws-lambda";
import fileParserService from "@libs/FileParserService";
import { getObjectKey } from "../../providers/s3client";
import sqsClient from "../../providers/sqsClient";

export const importFileParser: S3Handler = async (event) => {
  console.log("Parsing products file");

  const objectKey = getObjectKey(event);

  const s3ObjectReadableStream =
    fileParserService.getObjectReadableStream(objectKey);

  s3ObjectReadableStream
    .on("error", (e) => {
      console.error(e.message);
    })
    .on("end", () => {
      console.log("Complete reading the file");
    });

  console.log("Reading chunks from the file");
  for await (const chunk of s3ObjectReadableStream) {
    console.log("Sending product to queue", chunk);
    const sqsResponse = await sqsClient.sendMessage({ payload: chunk });
    console.log("Message have been sent", sqsResponse);
  }

  console.log("Moving to parsed folder");
  try {
    await fileParserService.moveToParsedFolder(objectKey);
  } catch (e) {
    console.error("Error while moving to parsed folder", e.message);
  }
};
