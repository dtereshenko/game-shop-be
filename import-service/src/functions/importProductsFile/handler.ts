import { APIGatewayHandler, formatJSONResponse } from "@libs/api-gateway";
import fileImportService from "@libs/FileImportService";
import {
  FILE_NAME_PARAM_NAME,
  schema,
  validatorOptions,
} from "@functions/importProductsFile/schema/validation";

export const importProductsFile: APIGatewayHandler = async (event) => {
  console.log("Importing products file");

  console.log("event", event);

  // Validate input parameters
  // '?? {}' is used to handle the case with event.queryStringParameters === undefined
  // since joi doesn't return error as expected in that case
  const { error } = schema.validate(
    event?.queryStringParameters ?? {},
    validatorOptions
  );

  console.log("error", error);
  if (error) {
    return formatJSONResponse({
      status: 400,
      response: { error: error?.details?.[0]?.message },
    });
  }

  const fileName = event.queryStringParameters[FILE_NAME_PARAM_NAME];

  try {
    const signedUrl = await fileImportService.getUploadUrl(fileName);
    console.log("signedUrl", signedUrl);

    return formatJSONResponse({ response: { url: signedUrl } });
  } catch (e) {
    console.log("Error while generating signed URL", e.message);

    return formatJSONResponse({
      status: 500,
      response: {
        message: e.message,
      },
    });
  }
};
