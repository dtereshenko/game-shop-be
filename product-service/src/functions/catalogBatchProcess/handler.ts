import { SQSHandler } from "aws-lambda";
import { getSnsClient } from "../../providers/snsClient";
// import productService from "@services/productService";

const snsCreateProductTopic = process.env.createProductSnsTopic;

export const catalogBatchProcess: SQSHandler = async (event) => {
  console.log("Start processing event from SQS", event);
  try {
    const snsClient = getSnsClient();
    for (const productUploadEvent of event.Records) {
      const product = JSON.parse(productUploadEvent.body);
      console.log("Trying to send notification for ", product);

      // Commented since I don't have free tier right now, so disabled my RDS for right now
      // await productService.create(product);

      await snsClient
        .publish({
          Message: `Product created ${JSON.stringify(product)}`,
          TopicArn: snsCreateProductTopic,
        })
        .promise();
      console.log("Notification sent", product);
    }
  } catch (e) {
    console.error(e.message);
  }
};
