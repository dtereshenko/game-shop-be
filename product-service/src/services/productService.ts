import got from "got";
import { GAMES_RESOURCE_URL } from "@config/api";
import { Product } from "../types/Product";

class ProductService {
  async getAll(): Promise<Product[]> | never {
    try {
      return await got(GAMES_RESOURCE_URL).json();
    } catch (e) {
      console.log("Error while retrieving products");
      console.log(e?.response?.body);
      throw e?.response?.body;
    }
  }

  async findById(id: string): Promise<Product> | never {
    try {
      return await got(`${GAMES_RESOURCE_URL}/${id}`).json();
    } catch (e) {
      console.log("Error while retrieving product by id ", id);
      console.log(e?.response?.body);
      throw e?.response?.body;
    }
  }
}

export default new ProductService();
