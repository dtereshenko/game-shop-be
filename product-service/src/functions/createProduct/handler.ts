import {
  formatJSONResponse,
  ValidatedEventAPIGatewayProxyEvent,
} from "@libs/api-gateway";
import productService from "@services/productService";

export const createProduct: ValidatedEventAPIGatewayProxyEvent = async (
  event
) => {
  try {
    const product = JSON.parse(event.body);
    console.log("Creating product with params", product);

    const status = await productService.create(product);

    return formatJSONResponse({
      response: { status: status ? "success" : "failed" },
    });
  } catch (e) {
    return formatJSONResponse({
      status: 500,
      response: {
        message: e.message,
      },
    });
  }
};
