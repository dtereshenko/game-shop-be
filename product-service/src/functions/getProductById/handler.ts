import {
  formatJSONResponse,
  ValidatedEventAPIGatewayProxyEvent,
} from "@libs/api-gateway";
import productService from "../../services/productService";

export const getProductById: ValidatedEventAPIGatewayProxyEvent = async (
  event
) => {
  const productId =
    event?.pathParameters?.productId || "e0e8b6a5-13f8-47f6-847e-9b22f0b16479";

  try {
    console.log("Requesting product by uuid", productId);

    const product = await productService.findById(productId);

    if (!product) {
      return formatJSONResponse({
        status: 404,
        response: { message: "Product not found" },
      });
    }

    return formatJSONResponse({
      response: { product },
    });
  } catch (e) {
    return formatJSONResponse({
      status: 500,
      response: {
        message: e.message,
      },
    });
  }
};
