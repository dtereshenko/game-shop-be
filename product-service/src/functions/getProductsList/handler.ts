import {
  formatJSONResponse,
  ValidatedEventAPIGatewayProxyEvent,
} from "@libs/api-gateway";
import productService from "../../services/productService";

export const getProductsList: ValidatedEventAPIGatewayProxyEvent = async () => {
  try {
    const games = await productService.getAll();

    return formatJSONResponse({
      response: { products: games },
    });
  } catch (e) {
    console.error("error message:", e.message);
    return formatJSONResponse({ status: 500, response: { message: e } });
  }
};
