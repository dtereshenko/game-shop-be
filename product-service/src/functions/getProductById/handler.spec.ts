import { getProductById } from "./handler";
import productService from "@services/productService";

const mockProduct = {
  title: "Street Fighters III",
  price: 15,
  description: "Fighting",
};

jest.mock("@services/productService", () => ({
  __esModule: true,
  default: {
    findById: jest.fn(() => Promise.resolve(mockProduct)),
  },
}));

describe("getProductById handler", function () {
  it("should return response 200 with the requested product", async function () {
    const productResponse = await getProductById(null, null, () => {});

    // @ts-ignore
    const { statusCode, body } = productResponse;

    expect(statusCode).toEqual(200);
    expect(body).toMatchSnapshot();
  });

  it('should return response 404 with "Product not found" message', async function () {
    (productService.findById as jest.Mock).mockRejectedValue(
      "Product Not found"
    );

    const productResponse = await getProductById(null, null, () => {});

    // @ts-ignore
    const { statusCode, body } = productResponse;

    expect(statusCode).toEqual(404);
    expect(body).toMatchSnapshot();
  });
});
