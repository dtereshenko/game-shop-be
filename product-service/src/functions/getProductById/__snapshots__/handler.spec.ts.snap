// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProductById handler should return response 200 with the requested product 1`] = `"{\\"product\\":{\\"title\\":\\"Street Fighters III\\",\\"price\\":15,\\"description\\":\\"Fighting\\"}}"`;

exports[`getProductById handler should return response 404 with "Product not found" message 1`] = `"{\\"message\\":\\"Product not found\\"}"`;
